File: flutter/elvish_mail/lib/mime/mail_compose_mime.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

/// RFC5322 + PGP/MIME builders (parity with `static/mail/compose.jsx`).
class MailComposeMime {
  static String rejectHeaderInjection(String value, String label) {
    if (value.contains(RegExp(r'[\r\n\x00]'))) {
      throw MailComposeException('$label contains invalid characters');
    }
    return value;
  }

  static List<String> splitAddressList(String value) {
    final raw = value.replaceAll('\r\n', '\n').replaceAll('\n', ' ').trim();
    if (raw.isEmpty) {
      return [];
    }
    var depth = 0;
    var cur = '';
    final parts = <String>[];
    for (var i = 0; i < raw.length; i++) {
      final c = raw[i];
      if (c == '<') {
        depth += 1;
      } else if (c == '>') {
        depth = depth > 0 ? depth - 1 : 0;
      }
      if ((c == ',' || c == ';') && depth == 0) {
        final t = cur.replaceAll(RegExp(r'\s+'), ' ').trim();
        if (t.isNotEmpty) {
          parts.add(t);
        }
        cur = '';
        continue;
      }
      cur += c;
    }
    final t = cur.replaceAll(RegExp(r'\s+'), ' ').trim();
    if (t.isNotEmpty) {
      parts.add(t);
    }
    return parts;
  }

  static String canonicalEmailToken(String display) {
    final raw = display.trim();
    if (raw.isEmpty) {
      return '';
    }
    final bracket = RegExp(r'<\s*([^<>\s@]+@[^<>\s@]+)\s*>').firstMatch(raw);
    if (bracket != null) {
      return bracket.group(1)!.trim().toLowerCase();
    }
    final bare = RegExp(r'([A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,})', caseSensitive: false).firstMatch(raw);
    if (bare != null) {
      return bare.group(1)!.trim().toLowerCase();
    }
    return raw.toLowerCase();
  }

  static Map<String, dynamic> buildHeaderStub({
    required String from,
    required List<String> to,
    List<String> cc = const [],
    List<String> bcc = const [],
    required String subject,
    String inReplyTo = '',
    String references = '',
  }) {
    final out = <String, dynamic>{
      'subject': subject,
      'from': from,
      'to': to,
    };
    if (cc.isNotEmpty) {
      out['cc'] = cc;
    }
    if (bcc.isNotEmpty) {
      out['bcc'] = bcc;
    }
    final irt = inReplyTo.trim();
    final refs = references.trim();
    if (irt.isNotEmpty) {
      out['in_reply_to'] = irt;
    }
    if (refs.isNotEmpty) {
      out['references'] = refs;
    }
    return out;
  }

  static Uint8List buildRfc5322({
    required String from,
    required List<String> to,
    required String subject,
    required String body,
    List<String> cc = const [],
    List<String> bcc = const [],
    String inReplyTo = '',
    String references = '',
  }) {
    final safeFrom = rejectHeaderInjection(from.isEmpty ? 'anonymous' : from, 'From');
    final safeSubject = rejectHeaderInjection(subject.isEmpty ? '(no subject)' : subject, 'Subject');
    final ccJoin = cc.where((a) => a.isNotEmpty).map((a) => rejectHeaderInjection(a, 'Cc')).join(', ');
    final bccJoin = bcc.where((a) => a.isNotEmpty).map((a) => rejectHeaderInjection(a, 'Bcc')).join(', ');
    final irt = rejectHeaderInjection(inReplyTo.trim(), 'In-Reply-To');
    final refs = rejectHeaderInjection(references.trim(), 'References');

    final lines = <String>[
      'Date: ${_rfc822Date()}',
      'From: $safeFrom',
      'To: ${to.map((a) => rejectHeaderInjection(a, 'To')).join(', ')}',
    ];
    if (ccJoin.isNotEmpty) {
      lines.add('Cc: $ccJoin');
    }
    if (bccJoin.isNotEmpty) {
      lines.add('Bcc: $bccJoin');
    }
    lines.add('Subject: $safeSubject');
    if (irt.isNotEmpty) {
      lines.add('In-Reply-To: $irt');
    }
    if (refs.isNotEmpty) {
      lines.add('References: $refs');
    }
    lines.addAll([
      'Content-Type: text/plain; charset="utf-8"; protected-headers="v1"',
      'Content-Transfer-Encoding: 8bit',
      'MIME-Version: 1.0',
      '',
      body,
    ]);
    return Uint8List.fromList(utf8.encode(lines.join('\r\n')));
  }

  static Uint8List buildPgpMimeMessage({
    required String from,
    required List<String> to,
    List<String> cc = const [],
    required String subject,
    required String armoredCiphertext,
    String inReplyTo = '',
    String references = '',
  }) {
    final boundary = '=_elvish_${_randomHex(16)}';
    final domain = _extractDomain(from);
    final messageId = '${DateTime.now().millisecondsSinceEpoch}.${_randomHex(16)}@$domain';
    final normalized = _normalizeCrlf(armoredCiphertext).replaceAll(RegExp(r'\r\n$'), '');
    final ccList = cc.where((a) => a.isNotEmpty).toList();
    final irt = inReplyTo.trim();
    final refs = references.trim();

    final lines = <String>[
      'Message-ID: <$messageId>',
      'Date: ${_rfc822Date()}',
      'From: ${rejectHeaderInjection(from.isEmpty ? 'anonymous' : from, 'From')}',
      'To: ${to.join(', ')}',
    ];
    if (ccList.isNotEmpty) {
      lines.add('Cc: ${ccList.join(', ')}');
    }
    if (subject.isNotEmpty) {
      lines.add('Subject: ${rejectHeaderInjection(subject, 'Subject')}');
    }
    if (irt.isNotEmpty) {
      lines.add('In-Reply-To: $irt');
    }
    if (refs.isNotEmpty) {
      lines.add('References: $refs');
    }
    lines.addAll([
      'Content-Type: multipart/encrypted; protocol="application/pgp-encrypted"; boundary="$boundary"',
      'MIME-Version: 1.0',
      '',
      'This is an OpenPGP/MIME encrypted message.',
      '--$boundary',
      'Content-Type: application/pgp-encrypted',
      'Content-Description: PGP/MIME version identification',
      '',
      'Version: 1',
      '--$boundary',
      'Content-Type: application/octet-stream; name="encrypted.asc"',
      'Content-Disposition: inline; filename="encrypted.asc"',
      'Content-Transfer-Encoding: 7bit',
      '',
      normalized,
      '--$boundary--',
      '',
    ]);
    return Uint8List.fromList(utf8.encode(lines.join('\r\n')));
  }

  static String _normalizeCrlf(String text) {
    return text.replaceAll('\r\n', '\n').replaceAll('\r', '\n').replaceAll('\n', '\r\n');
  }

  static String _randomHex(int bytes) {
    final rnd = Random.secure();
    return List.generate(bytes, (_) => rnd.nextInt(256))
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();
  }

  static String _rfc822Date() {
    final d = DateTime.now().toUtc();
    const days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    return '${days[d.weekday - 1]}, ${d.day.toString().padLeft(2, '0')} ${months[d.month - 1]} ${d.year} '
        '${d.hour.toString().padLeft(2, '0')}:${d.minute.toString().padLeft(2, '0')}:${d.second.toString().padLeft(2, '0')} GMT';
  }

  static String _extractDomain(String from) {
    final raw = from.trim();
    final m = RegExp(r'@([^<>\s@]+)').firstMatch(raw);
    if (m != null) {
      return m.group(1)!.toLowerCase();
    }
    return 'elvish.local';
  }
}

class MailComposeException implements Exception {
  MailComposeException(this.message);
  final String message;
  @override
  String toString() => message;
}

File: IOS/IOS/Mail/MailComposeMime.swift
import Foundation

/// RFC5322 + PGP/MIME builders for compose (parity with `static/mail/compose.jsx`).
enum MailComposeMime {
    struct HeaderStub: Encodable, Sendable {
        var subject: String
        var from: String
        var to: [String]
        var cc: [String]?
        var bcc: [String]?
        var inReplyTo: String?
        var references: String?

        enum CodingKeys: String, CodingKey {
            case subject, from, to, cc, bcc
            case inReplyTo = "in_reply_to"
            case references
        }
    }

    static func rejectHeaderInjection(_ value: String, label: String) throws -> String {
        let s = value
        if s.contains("\r") || s.contains("\n") || s.contains("\u{0}") {
            throw MailComposeError.headerInjection(label)
        }
        return s
    }

    private static func sanitizeAddressList(_ addrs: [String], label: String) throws -> [String] {
        var out: [String] = []
        for addr in addrs where !addr.isEmpty {
            out.append(try rejectHeaderInjection(addr, label: label))
        }
        return out
    }

    static func splitAddressList(_ value: String) -> [String] {
        let raw = value.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\n", with: " ").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty else { return [] }
        var depth = 0
        var cur = ""
        var parts: [String] = []
        for ch in raw {
            if ch == "<" { depth += 1 }
            else if ch == ">" { depth = max(0, depth - 1) }
            if (ch == "," || ch == ";") && depth == 0 {
                let t = cur.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression).trimmingCharacters(in: .whitespacesAndNewlines)
                if !t.isEmpty { parts.append(t) }
                cur = ""
                continue
            }
            cur.append(ch)
        }
        let t = cur.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression).trimmingCharacters(in: .whitespacesAndNewlines)
        if !t.isEmpty { parts.append(t) }
        return parts
    }

    static func canonicalEmailToken(_ display: String) -> String {
        let raw = display.trimmingCharacters(in: .whitespacesAndNewlines)
        if raw.isEmpty { return "" }
        if let r = raw.range(of: #"<([^<>\s@]+@[^<>\s@]+)>"#, options: .regularExpression) {
            let inner = String(raw[r]).trimmingCharacters(in: CharacterSet(charactersIn: "<>"))
            return inner.lowercased()
        }
        if let r = raw.range(of: #"([A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,})"#, options: [.regularExpression, .caseInsensitive]) {
            return String(raw[r]).lowercased()
        }
        return raw.lowercased()
    }

    static func buildHeaderStub(
        from: String,
        to: [String],
        cc: [String],
        bcc: [String],
        subject: String,
        inReplyTo: String = "",
        references: String = ""
    ) throws -> HeaderStub {
        var stub = HeaderStub(
            subject: subject,
            from: from,
            to: to,
            cc: cc.isEmpty ? nil : cc,
            bcc: bcc.isEmpty ? nil : bcc
        )
        let irt = inReplyTo.trimmingCharacters(in: .whitespacesAndNewlines)
        let refs = references.trimmingCharacters(in: .whitespacesAndNewlines)
        if !irt.isEmpty { stub.inReplyTo = irt }
        if !refs.isEmpty { stub.references = refs }
        return stub
    }

    static func buildRFC5322(
        from: String,
        to: [String],
        subject: String,
        body: String,
        cc: [String] = [],
        bcc: [String] = [],
        inReplyTo: String = "",
        references: String = ""
    ) throws -> Data {
        let safeFrom = try rejectHeaderInjection(from.isEmpty ? "anonymous" : from, label: "From")
        let safeSubject = try rejectHeaderInjection(subject.isEmpty ? "(no subject)" : subject, label: "Subject")
        let safeTo = try sanitizeAddressList(to, label: "To")
        let ccJoin = try sanitizeAddressList(cc, label: "Cc").joined(separator: ", ")
        let bccJoin = try sanitizeAddressList(bcc, label: "Bcc").joined(separator: ", ")
        let irt = try rejectHeaderInjection(inReplyTo.trimmingCharacters(in: .whitespacesAndNewlines), label: "In-Reply-To")
        let refs = try rejectHeaderInjection(references.trimmingCharacters(in: .whitespacesAndNewlines), label: "References")

        var lines: [String] = []
        lines.append("Date: \(rfc822Date())")
        lines.append("From: \(safeFrom)")
        lines.append("To: \(safeTo.joined(separator: ", "))")
        if !ccJoin.isEmpty { lines.append("Cc: \(ccJoin)") }
        if !bccJoin.isEmpty { lines.append("Bcc: \(bccJoin)") }
        lines.append("Subject: \(safeSubject)")
        if !irt.isEmpty { lines.append("In-Reply-To: \(irt)") }
        if !refs.isEmpty { lines.append("References: \(refs)") }
        lines.append("Content-Type: text/plain; charset=\"utf-8\"; protected-headers=\"v1\"")
        lines.append("Content-Transfer-Encoding: 8bit")
        lines.append("MIME-Version: 1.0")
        lines.append("")
        lines.append(body)
        return Data(lines.joined(separator: "\r\n").utf8)
    }

    static func buildPGPMIMEMessage(
        from: String,
        to: [String],
        cc: [String],
        subject: String,
        armoredCiphertext: String,
        inReplyTo: String = "",
        references: String = ""
    ) throws -> Data {
        let boundary = "=_elvish_\(randomHex(16))"
        let domain = extractDomain(from: from)
        let messageID = "\(Int(Date().timeIntervalSince1970 * 1000)).\(randomHex(16))@\(domain)"
        let normalized = normalizeCRLF(armoredCiphertext).trimmingCharacters(in: CharacterSet(charactersIn: "\r\n"))
        let ccList = cc.filter { !$0.isEmpty }
        let irt = inReplyTo.trimmingCharacters(in: .whitespacesAndNewlines)
        let refs = references.trimmingCharacters(in: .whitespacesAndNewlines)

        var lines: [String] = [
            "Message-ID: <\(messageID)>",
            "Date: \(rfc822Date())",
            "From: \(try rejectHeaderInjection(from.isEmpty ? "anonymous" : from, label: "From"))",
            "To: \(to.joined(separator: ", "))",
        ]
        if !ccList.isEmpty { lines.append("Cc: \(ccList.joined(separator: ", "))") }
        if !subject.isEmpty { lines.append("Subject: \(try rejectHeaderInjection(subject, label: "Subject"))") }
        if !irt.isEmpty { lines.append("In-Reply-To: \(irt)") }
        if !refs.isEmpty { lines.append("References: \(refs)") }
        lines.append("Content-Type: multipart/encrypted; protocol=\"application/pgp-encrypted\"; boundary=\"\(boundary)\"")
        lines.append("MIME-Version: 1.0")
        lines.append("")
        lines.append("This is an OpenPGP/MIME encrypted message.")
        lines.append("--\(boundary)")
        lines.append("Content-Type: application/pgp-encrypted")
        lines.append("Content-Description: PGP/MIME version identification")
        lines.append("")
        lines.append("Version: 1")
        lines.append("--\(boundary)")
        lines.append("Content-Type: application/octet-stream; name=\"encrypted.asc\"")
        lines.append("Content-Disposition: inline; filename=\"encrypted.asc\"")
        lines.append("Content-Transfer-Encoding: 7bit")
        lines.append("")
        lines.append(normalized)
        lines.append("--\(boundary)--")
        lines.append("")
        return Data(lines.joined(separator: "\r\n").utf8)
    }

    private static func normalizeCRLF(_ text: String) -> String {
        text.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
            .replacingOccurrences(of: "\n", with: "\r\n")
    }

    private static func randomHex(_ bytes: Int) -> String {
        let data = (try? ElvishAccountWrap.randomBytes(bytes)) ?? Data((0 ..< bytes).map { _ in UInt8.random(in: 0 ... 255) })
        return data.map { String(format: "%02x", $0) }.joined()
    }

    private static func rfc822Date() -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        return formatter.string(from: Date())
    }

    private static func extractDomain(from: String) -> String {
        let raw = from.trimmingCharacters(in: .whitespacesAndNewlines)
        if let r = raw.range(of: #"@([^<>\s@]+)"#, options: .regularExpression) {
            let part = String(raw[r]).replacingOccurrences(of: "@", with: "")
            return part.lowercased()
        }
        return "elvish.local"
    }
}

enum MailComposeError: Error, LocalizedError {
    case headerInjection(String)

    var errorDescription: String? {
        switch self {
        case let .headerInjection(label):
            return "\(label) contains invalid characters"
        }
    }
}
